hudi-utilities/src/test/java/org/apache/hudi/utilities/TestMultiFormatTimestampBasedKeyGenerator.java

```diff
@@ -0,0 +1,206 @@
+    }
+    if (outputTimezone != null) {
+      properties.setProperty("hoodie.deltastreamer.keygen.timebased.output.timezone", outputTimezone);
+    }
+
+    return properties;
+  }
+
+  @Test
+  public void test_ExpectsMatch_SingleInputFormat_ISO8601WithMsZ_OutputTimezoneAsUTC() {
+    baseRecord.put("createTime", "2020-04-01T13:01:33.428Z");
+    properties = this.getBaseKeyConfig(
+            "DATE_STRING",
+            "yyyy-MM-dd'T'HH:mm:ss.SSSZ",
+            "",
+            "",
+            "yyyyMMddHH",
+            "UTC");
+    HoodieKey hk1 = new MultiFormatTimestampBasedKeyGenerator(properties).getKey(baseRecord);
+
+    assertEquals("2020040113", hk1.getPartitionPath());
+  }
+
+  @Test
+  public void test_ExpectsMatch_SingleInputFormats_ISO8601WithMsZ_OutputTimezoneAsInputDateTimeZone() {
+    baseRecord.put("createTime", "2020-04-01T13:01:33.428Z");
+    properties = this.getBaseKeyConfig(
+            "DATE_STRING",
+            "yyyy-MM-dd'T'HH:mm:ss.SSSZ",
+            "",
+            "",
+            "yyyyMMddHH",
+            "");
+    HoodieKey hk1 = new MultiFormatTimestampBasedKeyGenerator(properties).getKey(baseRecord);
+
+    assertEquals("2020040113", hk1.getPartitionPath());
+  }
+
+  @Test
+  public void test_ExpectsMatch_MultipleInputFormats_ISO8601WithMsZ_OutputTimezoneAsUTC() {
+    baseRecord.put("createTime", "2020-04-01T13:01:33.428Z");
+    properties = this.getBaseKeyConfig(
+            "DATE_STRING",
+            "yyyy-MM-dd'T'HH:mm:ssZ,yyyy-MM-dd'T'HH:mm:ss.SSSZ",
+            "",
+            "",
+            "yyyyMMddHH",
+            "UTC");
+    HoodieKey hk1 = new MultiFormatTimestampBasedKeyGenerator(properties).getKey(baseRecord);
+
+    assertEquals("2020040113", hk1.getPartitionPath());
+  }
+
+  @Test
+  public void test_ExpectsMatch_MultipleInputFormats_ISO8601NoMsZ_OutputTimezoneAsUTC() {
+    baseRecord.put("createTime", "2020-04-01T13:01:33Z");
+    properties = this.getBaseKeyConfig(
+            "DATE_STRING",
+            "yyyy-MM-dd'T'HH:mm:ssZ,yyyy-MM-dd'T'HH:mm:ss.SSSZ",
+            "",
+            "",
+            "yyyyMMddHH",
+            "UTC");
+    HoodieKey hk1 = new MultiFormatTimestampBasedKeyGenerator(properties).getKey(baseRecord);
+
+    assertEquals("2020040113", hk1.getPartitionPath());
+  }
+
+  @Test
+  public void test_ExpectsMatch_MultipleInputFormats_ISO8601NoMsWithOffset_OutputTimezoneAsUTC() {
+    baseRecord.put("createTime", "2020-04-01T13:01:33-05:00");
+    properties = this.getBaseKeyConfig(
+            "DATE_STRING",
+            "yyyy-MM-dd'T'HH:mm:ssZ,yyyy-MM-dd'T'HH:mm:ss.SSSZ",
+            "",
+            "",
+            "yyyyMMddHH",
+            "UTC");
+    HoodieKey hk1 = new MultiFormatTimestampBasedKeyGenerator(properties).getKey(baseRecord);
+
+    assertEquals("2020040118", hk1.getPartitionPath());
+  }
+
+  @Test
+  public void test_ExpectsMatch_MultipleInputFormats_ISO8601WithMsWithOffset_OutputTimezoneAsUTC() {
```

Comment:
Wondering if we can combine test cases here by having one more GenericRecord variable defined in this class?

Reply:
Yup, I can look into doing that as well.